langserver/handlers/did_change.go

```diff
@@ -2,26 +2,74 @@ package handlers
 
 import (
 	"context"
+	"encoding/json"
+	"fmt"
 
 	lsctx "github.com/hashicorp/terraform-ls/internal/context"
 	ilsp "github.com/hashicorp/terraform-ls/internal/lsp"
 	lsp "github.com/sourcegraph/go-lsp"
 )
 
-func TextDocumentDidChange(ctx context.Context, params lsp.DidChangeTextDocumentParams) error {
+func TextDocumentDidChange(ctx context.Context, params DidChangeTextDocumentParams) error {
+	p := lsp.DidChangeTextDocumentParams{
+		TextDocument: lsp.VersionedTextDocumentIdentifier{
+			TextDocumentIdentifier: lsp.TextDocumentIdentifier{
+				URI: params.TextDocument.URI,
+			},
+			Version: params.TextDocument.Version,
+		},
+		ContentChanges: params.ContentChanges,
+	}
+
 	fs, err := lsctx.Filesystem(ctx)
 	if err != nil {
 		return err
 	}
 
-	fh := ilsp.VersionedFileHandler(params.TextDocument)
+	fh := ilsp.VersionedFileHandler(p.TextDocument)
 	f, err := fs.GetFile(fh)
 	if err != nil {
 		return err
 	}
+
+	// old version change, just skip
+	if p.TextDocument.Version <= f.Version() {
+		return fmt.Errorf("skip old version %d change: %#v, current file version %d", p.TextDocument.Version, params, f.Version())
+	}
+
+	// missing version
+	if p.TextDocument.Version > f.Version()+1 {
+		return fmt.Errorf("missing version %d change: %#v, current file version %d", p.TextDocument.Version, params, f.Version())
+	}
```

Comment:
It may sound like a nitpick, but if we do this to surface client-side bugs, I think we should make the messaging a bit more actionable and human-friendly and we should be marking the virtual file as closed, so that we avoid completing anything on outdated documents

```suggestion
	if p.TextDocument.Version <= f.Version() {
		fs.Close(fh)
		return fmt.Errorf("Old version (%d) received, current version is %d. "+
			"Unable to update %s. This is likely a bug, please report it.",
			p.TextDocument.Version, f.Version(), p.TextDocument.URI)
	}

	if p.TextDocument.Version > f.Version()+1 {
		fs.Close(fh)
		return fmt.Errorf("New version (%d) received, current version is %d. "+
			"Unable to update %s. This is likely a bug, please report it.",
			p.TextDocument.Version, f.Version(), p.TextDocument.URI)
	}
```

Side note: Because `didChange` is actually a notification, clients won't surface these errors anywhere (as per LSP spec that notification responses are ignored by clients), so we'd only see this in a log file, but I'm happy to address that in a separate PR - we need to look into [diagnostic publishing](https://github.com/hashicorp/terraform-ls/issues/27) anyway.

Reply:
Yes, I think another way could be implementing the `window/showMessage` handler to show messages to user, then let them close and reopen that file

Reply:
Yes - totally, but we can address that in a separate followup PR.